### DISCUSSION TITLE
append metadata to hs.window instance

### DISCUSSION THREAD

**Discussion by mogenson** [Hammerspoon/hammerspoon#3386] [Q&A]
Hello, I'd like to append some metadata to an instance of an `hs.window` so I can retrieve and inspect it later. For example, use the `hs.window:newWatcher()` method to create a `uielement.watcher` and store it with the window.

```lua
myWindow = hs.window.focusedWindow()
myWindow.myWatcher = myWindow:newWatcher( -- can't index an hs.window value
        function(element, event, watcher)
            print("myWatcher for myWindow called with event " .. event)
        end
    )
myWindow.myWatcher:start({hs.uielement.watcher.windowMoved})
```

Are there any strategies in Lua to storing Lua variables/tables alongside a `userdata` type like `hs.window`? Wold `debug. setuservalue()` be an option? I can't find any good examples of it's use.

**Comment by Rhys-T** [comment]
I don't think there's any way to add that info to the `userdata` itself. You're better off maintaining any information like that in a table. (`debug.setuservalue` seems to be limited to messing with the `userdata`'s existing fields, and can't add new ones.)

Using the `hs.window` instance as a table key directly won't work like you expect, though: Each time Hammerspoon gives you a reference to a window, it's a _new_ `userdata`, even if it's pointing to the same actual window. They compare as equal using the `==` operator, but they aren't `rawequal` to each other, and they count as different table keys.[^otherdups] What you can do instead is get each window's [`:id()`][] and use _that_ as an index into your table. Something like this:

```lua
-- somewhere global:
myWindowWatchers = {}

-- wherever you're creating each watcher:
myWindow = hs.window.focusedWindow()
myWindowID = myWindow:id()
myWindowWatchers[myWindowID] = myWindow:newWatcher(
        function(element, event, watcher)
            print("myWatcher for myWindow called with event " .. event)
            -- clean up watchers for windows that no longer exist:
            if event == hs.uielement.watcher.elementDestroyed then
                myWindowWatchers[myWindowID] = nil
            end
        end
    )
myWindow.myWatcher:start({hs.uielement.watcher.windowMoved, hs.uielement.watcher.elementDestroyed})
-- make sure that elementDestroyed gets added to this call: ↑
```

[^otherdups]: See #2720, #3532, and #3057 for more instances of that sort of thing in Hammerspoon's other `userdata`-based types. This is another reason why you don't really want to attach extra fields directly to the `userdata` - they won't be there on the _next_ `userdata` for that window!

[`:id()`]: https://www.hammerspoon.org/docs/hs.window.html#id

**Comment by mogenson** [comment]
Thanks for the response. I ended up using window IDs as keys for some locally managed tables here: https://github.com/mogenson/PaperWM.spoon/blob/e64e0d864d2db4cb6f65a92929bbcdf4a1b07fa4/init.lua#L121
